New feature: Rename pde to fit with Matlab DG-SparseGrid

hbrunie opened the pull request:
## Proposed changes

This PR is here to solve [the rename pde issue#344](https://github.com/project-asgard/asgard/issues/344).
It involves adding a parameter to asgard binary.

```
  --case_selection <e.g. 1 (or 2)>      PDE case to solve; default is 1,
                                            sometimes there is a second case
                                            possible.
```

## What type(s) of changes does this code introduce?
_Put an `x` in the boxes that apply._

- [ ] Bugfix
- [x] New feature
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Documentation content changes
- [ ] Other (please describe):

### Does this introduce a breaking change?

- [ ] Yes
- [x] No

## What systems has this change been tested on?

## Checklist

- [x] this PR is up to date with current the current state of 'develop'
- [x] code added or changed in the PR has been clang-formatted
- [x] this PR adds tests to cover any new code, or to catch a bug that is being fixed
- [ ] documentation has been added (if appropriate)

hbrunie committed:
Rename pde fokkerplanck 4p1a and 4p2

fokkerplanck1 from 4p1a to pitch E and
fokkerplanck1 from 4p2 to pitch C and
Change all 4p1a by pitch_E (into testing coefficients and time_advance
too).
clang format.
Fix a non-initialized variable warning in lib_dispatch.

hbrunie committed:
Adding fokkerplanck pitch E case 2.

To add case we are using one more template value for PDE.
TODO: add case2 proper gold generated tests for coefficients and
time advance.

hbrunie committed:
Use macro to wrap assert (macro).

This fix the bug of not having real error root when assertion fails.

hbrunie commented:
I have rebased this PR above the branch of @bmcdanie fixing the time in analytic solution generation.
Once @bmcdanie PR would have been merged to asgard/develop I will have to rebase on develop and force push again.

hbrunie commented:
To pass the tests `coefficients` and `time_advance`, we need the gold generated corresponding files for `fokkerplanck1_pitch_E_case2` (gaussian) from the Matlab `DG-SparseGrid` repository.

This is not implemented yet. This is the corresponding [PR] (https://github.com/project-asgard/DG-SparseGrid/pull/65) from this other repo.

hbrunie committed:
Fix fokkerplanck 1d case2 tests.

bmcdanie commented on src/program_options.hpp:
```diff
@@ -250,6 +267,7 @@ class parser
   std::string pde_str = DEFAULT_PDE_STR;
   // pde to construct/evaluate
   PDE_opts pde_choice = DEFAULT_PDE_OPT;
+  // pde selected case (f0)
```
what does this comment refer to?

hbrunie replied:
I forgot to remove ..

bmcdanie commented on src/program_options.hpp:
```diff
@@ -111,18 +127,19 @@ class parser
+  static auto constexpr DEFAULT_PDE_STR           = "continuity_2";
+  static auto constexpr DEFAULT_PDE_OPT           = PDE_opts::continuity_2;
+  static auto constexpr DEFAULT_SOLVER            = solve_opts::direct;
+  static auto constexpr DEFAULT_PDE_SELECTED_CASE = 0;
```
can you init this with the enum value instead of an integer?

hbrunie replied:
Yes.

bmcdanie commented on src/tensors.hpp:
```diff
@@ -150,13 +150,13 @@ class vector
   vector(vector<P, mem, resrc> &&);
 
-  // as with copy assignment, static tools::expect added
+  // as with copy assignment, static expect added
```
I believe this is my mistake originally - should say "static assert" but my sed changes to "tools::expect", then your sed changed that to "static expect".

hbrunie replied:
OK.

hbrunie replied:
I also changed the one line 153.

bmcdanie commented on src/tensors.hpp:
```diff
@@ -150,13 +150,13 @@ class vector
   // cannot be templated per C++ spec 12.8
   // instead of disabling w/ sfinae for const_view,
-  // static tools::expect added to definition
+  // static expect added to definition
```
"static expect" -> "static assert"

hbrunie replied:
Yes

bmcdanie commented on src/pde/pde_fokkerplanck1_pitch_E.hpp:
```diff
@@ -68,19 +68,24 @@ class PDE_fokkerplanck_1d_4p1a : public PDE<P>
+      P t1 = 1 - std::pow(p, 2);
+      P t2 = 1 - std::pow(z(i), 2);
+      P t3 = 0.;
+      if constexpr (user_case == PDE_case_opts::mod1)
```
we have a convention that we always use curlies around conditionals, even single line ones

hbrunie replied:
yes

bmcdanie reviewed (changes requested):
thanks for this PR! couple of minor notes, looks great overall.

hbrunie committed:
Requested changes to PR.

Remove obsolete comment.
Enum instead of int.
static expect -> static assert
brackets for one line if/else.
